abaco.go

```diff
@@ -388,17 +406,171 @@ func enumerateAbacoRings() (rings []int, err error) {
 
 //------------------------------------------------------------------------------------------------
 
-// AbacoSource represents all AbacoRing ring buffers that can potentially supply data, as well
-// as all AbacoGroups that are discovered in the SampleData phase.
+// AbacoUDPReceiver represents a single Abaco device producing data by UDP packets to a single UDP port.
+type AbacoUDPReceiver struct {
+	host     string       // in the form: "127.0.0.1:56789"
+	conn     *net.UDPConn // active UDP connection
+	data     chan []*packets.Packet
+	sendmore chan bool
+}
+
+// NewAbacoUDPReceiver creates a new AbacoUDPReceiver and binds as a server to the requested host:port
+func NewAbacoUDPReceiver(hostport string) (dev *AbacoUDPReceiver, err error) {
+	dev = new(AbacoUDPReceiver)
+	dev.host = hostport
+	if _, err := net.ResolveUDPAddr("udp", hostport); err != nil {
+		return nil, err
+	}
+	return dev, nil
+}
+
+// start starts the UDP device by...?
+func (device *AbacoUDPReceiver) start() (err error) {
+	raddr, err := net.ResolveUDPAddr("udp", device.host)
+	if err != nil {
+		return err
+	}
+	conn, err := net.ListenUDP("udp", raddr)
+	if err != nil {
+		return err
+	}
+	const UDPPacketSize = 8192
+	bufferPool := sync.Pool{
+		New: func() interface{} { return make([]byte, UDPPacketSize) },
+	}
+	device.conn = conn
+
+	device.sendmore = make(chan bool)
+	device.data = make(chan []*packets.Packet)
+	singlepackets := make(chan []byte, 20)
+	// This goroutine handles UDP message sent one at a time on singlepackets by the other goroutine.
+	// It converts them into packet.Packet objects, queues them into a slice of *packets.Packet
+	// pointers and sends the whole slice when a request comes on device.sendmore.
+	go func() {
+		defer close(device.data)
+		defer func() { device.conn = nil }()
+
+		const initialQueueCapacity = 20000
+		queue := make([]*packets.Packet, 0, initialQueueCapacity)
```

Comment:
This feels like a potential performance trap. We're allocating space for 20k packets repeatedly? Are we potentially only using a small fraction of that space, and thus using up much more memory than we need? What is the purpose of this sort of double buffer, why can't whatever comes next just read from single packets?

Reply:
No, we're allocating space for 20k _pointers_ (to packets) repeatedly. That's going to be 160,000 bytes. We expect up to several x 10,000 packets per second. The queue will have to hold some ~50 ms of data before we pass it to the handler, which could be several thousand packets. By sizing it to a capacity of 20k at the start, we can prevent the reallocate-and-copy steps required each time it has to grow.

That "~50 ms" comes from line 818 in this function: https://github.com/usnistgov/dastard/blob/8d3358c78bc6d1f909d1bc8c2cd4c18d042dc9cc/abaco.go#L811-L821

Anyway, if we are running 250 MB/sec and 8100 payload bytes per UDP packet, then we expect only about 1500 packets per readout. Probably, then, we should change this initial capacity to 2048. If it so happens that the rate is higher, or the packets have only half the payload (as in one recent test), then maybe the queue would have to resize once, but only once.

This is a performance tuning. I believe we want this capacity to be bigger than the usual number of packets expected per readout, but the costs are small if we guess low OR high. I'll change 20000 -> 2048, but it's probably a small matter.

Reply:
Ok, I guess I still don't see why we do this at all, but it's at least not much memory so should be fine.